### PR TITLE
[JUJU-1616] Add backend for secrets access control

### PR DESCRIPTION
**Pull request by wallyworld**
This PR implements the secret ACL backend. The stub permission checks on the secrets manager facade were reworked to suit. It also fixes a few issues with the grant/revoke hook commands that appeared when the full QA was done.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- [X] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

deploy test charms
```
juju deploy juju-qa-dummy-source
juju deploy juju-qa-dummy-sink
juju add-unit dummy-sink
```
create a secret
```
juju exec --unit dummy-source/0 "secret-add data=foo"
secret:cbtcu5efg3o24llgfr7g
```
create check the owner can update and read a secret
```
juju exec --unit dummy-source/0 "secret-update secret:cbtcu5efg3o24llgfr7g data=foo2"
juju exec --unit dummy-source/0 "secret-get secret:cbtcu5efg3o24llgfr7g"
data: foo2
```
check that unauthorised access not allowed
```
juju exec --unit dummy-sink/0 "secret-get secret:cbtcu5efg3o24llgfr7g"
ERROR permission denied
```
relate dummy-sourc to dummy-sink and grant access
```
juju relate dummy-source dummy-sink
juju exec --unit dummy-source/0 "secret-grant secret:cbtcu5efg3o24llgfr7g -r 0"
```
any unit of dummy-sink can read the secret
```
juju exec --unit dummy-sink/0 "secret-get secret:cbtcu5efg3o24llgfr7g"
data: foo2
juju exec --unit dummy-sink/1 "secret-get secret:cbtcu5efg3o24llgfr7g"
data: foo2
```
remove the relation and access goes away
```
juju remove-relation dummy-source:sink dummy-sink:source
juju exec --unit dummy-sink/1 "secret-get secret:cbtcu5efg3o24llgfr7g"
ERROR permission denied
```
add relation again and grant access just to dummy-sink/1
```
juju relate dummy-source dummy-sink
juju exec --unit dummy-source/0 "secret-grant secret:cbtcu5efg3o24llgfr7g -r 1 --unit dummy-sink/1"
juju exec --unit dummy-sink/1 "secret-get secret:cbtcu5efg3o24llgfr7g"
data: foo2
juju exec --unit dummy-sink/0 "secret-get secret:cbtcu5efg3o24llgfr7g"
ERROR permission denied
```
revoke access and check
```
juju exec --unit dummy-source/0 "secret-revoke secret:cbtcu5efg3o24llgfr7g --unit dummy-sink/1"
uju exec --unit dummy-sink/1 "secret-get secret:cbtcu5efg3o24llgfr7g"
ERROR permission denied
```

### CODE REVIEW

File: state/secrets.go [hunk 1]
```diff
@@ -552,14 +554,191 @@ func (st *State) WatchConsumedSecretsChanges(consumer string) StringsWatcher {
...
+		docID = id
+	default:
+		err = errors.NotValidf("secret scope reference %q", tag.String())
+	}
+	return entity, collName, docID, err
+}
+
+// GrantSecretAccess saves the secret access role for the subject with the specified scope.
+func (st *State) GrantSecretAccess(uri *secrets.URI, scope names.Tag, subject names.Tag, role secrets.SecretRole) (err error) {
+	if role == "" || !role.IsValid() {
```
Comment (ycliuhw) [inline state/secrets.go:588]: If we think `RoleNone` is invalid then we just need `if !role.IsValid()`

File: state/secrets.go [hunk 2]
```diff
@@ -552,14 +554,191 @@ func (st *State) WatchConsumedSecretsChanges(consumer string) StringsWatcher {
...
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		n, err := secretMetadataCollection.FindId(uri.ID).Count()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if n == 0 {
+			return nil, jujutxn.ErrNoOperations
+		}
+		err = secretPermissionsCollection.FindId(docID).One(&doc)
+		if err := errors.Cause(err); err != nil && err != mgo.ErrNotFound {
```
Comment (ycliuhw) [inline state/secrets.go:678]:
```go
if err == mgo.ErrNotFound {
			return nil, jujutxn.ErrNoOperations
		}
if err := errors.Cause(err); err != nil {
			return nil, errors.Trace(err)
		}
		

```

**Review by ycliuhw — APPROVED** [review]
LGTM, just a few small issues. ty

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 3]
```diff
@@ -63,7 +55,8 @@ func (s *SecretsManagerAPI) createSecret(ctx context.Context, arg params.CreateS
 	if err != nil {
 		return "", errors.Trace(err)
 	}
-	if !canManage(secretOwner) {
+	// A unit can create a secret for its own app.
+	if authTagApp(s.authTag) != secretOwner.Id() {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/secretsmanager/secrets.go:59]: A secret is currently owned by an application. The unit leader of that app can create a secret owned by that app (not enforced yet). I'll reword the comment if it's unclear.

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 4]
```diff
@@ -75,6 +68,11 @@ func (s *SecretsManagerAPI) createSecret(ctx context.Context, arg params.CreateS
 	if err != nil {
 		return "", errors.Trace(err)
 	}
+	err = s.secretsConsumer.GrantSecretAccess(uri, secretOwner, secretOwner, coresecrets.RoleManage)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/secretsmanager/secrets.go:71]: yes, the charm which created the secret needs to be able to update/manage it

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 5]
```diff
@@ -75,6 +68,11 @@ func (s *SecretsManagerAPI) createSecret(ctx context.Context, arg params.CreateS
...
+	if err != nil {
+		// TODO(wallyworld) - remove secret when that is supported
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/secretsmanager/secrets.go:73]: we don't support deleting secrets yet
and if this grant fails, we don't want to leave the half created secret behiind

File: core/secrets/rbac.go [hunk 6]
```diff
@@ -20,3 +21,24 @@ func (r SecretRole) IsValid() bool {
 	}
 	return false
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply core/secrets/rbac.go:22]: yeah, should be

File: core/secrets/rbac.go [hunk 7]
```diff
@@ -20,3 +21,24 @@ func (r SecretRole) IsValid() bool {
...
+
+func (r SecretRole) value() int {
+	switch r {
+	case RoleView:
+		return 1
+	case RoleRotate:
+		return 2
+	case RoleManage:
+		return 3
+	default:
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply core/secrets/rbac.go:33]: -1

File: core/secrets/rbac_test.go [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+type RoleSuite struct{}
+
+var _ = gc.Suite(&RoleSuite{})
+
+func (s *SecretValueSuite) TestAllowed(c *gc.C) {
+	c.Assert(secrets.RoleView.Allowed(secrets.RoleView), jc.IsTrue)
+	c.Assert(secrets.RoleView.Allowed(secrets.RoleRotate), jc.IsFalse)
+	c.Assert(secrets.RoleView.Allowed(secrets.RoleManage), jc.IsFalse)
+	c.Assert(secrets.RoleRotate.Allowed(secrets.RoleView), jc.IsTrue)
+	c.Assert(secrets.RoleRotate.Allowed(secrets.RoleRotate), jc.IsTrue)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply core/secrets/rbac_test.go:22]: yup

File: state/secrets.go [hunk 9]
```diff
@@ -397,7 +397,9 @@ func secretConsumerKey(id, consumer string) string {
 func (st *State) checkConsumerCountOps(uri *secrets.URI, inc int) ([]txn.Op, error) {
 	refCountCollection, ccloser := st.db().GetCollection(refcountsC)
 	defer ccloser()
-	countOp, _, err := nsRefcounts.CurrentOp(refCountCollection, uri.ID)
+
+	key := fmt.Sprintf("%s#consumer", uri.ID)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply state/secrets.go:401]: and it's only used here in this one place

File: state/secrets.go (hunk 1 above)
Comment (wallyworld) [reply state/secrets.go:588]: it's not valid to grant role none, that is done using revoke.
so it's ok to disallow role none in grant operation
changed "" to RoleNone

File: state/secrets.go (hunk 2 above)
Comment (wallyworld) [reply state/secrets.go]: fixed, here and in 3 other places

**Pull request merged**